{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "13066e38",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "import seaborn as sns\n",
    "import re\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from statsmodels.formula.api import ols, logit\n",
    "import warnings\n",
    "from statsmodels.tools.sm_exceptions import ConvergenceWarning\n",
    "np.random.seed(0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb5984bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "Genes = pd.read_csv(\"ClusterDataset.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "57ac9b2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "Genes = Genes[Genes[\"Name\"]!=\"gene_variance\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bb9226e1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUBJID</th>\n",
       "      <th>SEX</th>\n",
       "      <th>AGE</th>\n",
       "      <th>DTHHRDY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GTEX-1117F</td>\n",
       "      <td>2</td>\n",
       "      <td>60-69</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       SUBJID  SEX    AGE  DTHHRDY\n",
       "0  GTEX-1117F    2  60-69      4.0"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Subjects = pd.read_csv(\"SubjectsMeta.txt\",sep='\\t')\n",
    "Subjects.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "528bde14",
   "metadata": {},
   "outputs": [],
   "source": [
    "#get all genecolumns \n",
    "gene_cols = [g for g in Genes.columns if g.startswith(\"ENSG\")]\n",
    "#Remove gene column and features column convert to numpy array\n",
    "gene_array = Genes.iloc[:,1:-1].to_numpy()\n",
    "norm_genes = StandardScaler().fit_transform(gene_array)\n",
    "norm_genes = pd.DataFrame(norm_genes,columns=gene_cols)\n",
    "norm_genes[\"Name\"]=Genes[\"Name\"]\n",
    "norm_genes[\"tissue\"]=Genes[\"tissue\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df1c15d5",
   "metadata": {},
   "source": [
    "### Merge Genes with Subject Meta Data\n",
    "\n",
    "Leave age as a categorical variable for correction in regression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "dabb06fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "#use regex to get subjid take GTEX followed by sep char then match 1 or more char that are not that sep\n",
    "pattern = r'^(GTEX\\x2D[^\\x2D]+)'\n",
    "norm_genes[\"SUBJID\"] = norm_genes[\"Name\"].apply(lambda x: re.match(pattern,x))\n",
    "#clean bad matches (need to drop variance row from preprocessing - failed to do this earlier)\n",
    "norm_genes=norm_genes[~(norm_genes.index==17382)]\n",
    "norm_genes[\"SUBJID\"] = norm_genes[\"SUBJID\"].apply(lambda x: x.group(1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3d4ce46e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSG00000244734.3</th>\n",
       "      <th>ENSG00000210082.2</th>\n",
       "      <th>ENSG00000198804.2</th>\n",
       "      <th>ENSG00000198712.1</th>\n",
       "      <th>ENSG00000198938.2</th>\n",
       "      <th>ENSG00000188536.12</th>\n",
       "      <th>ENSG00000198899.2</th>\n",
       "      <th>ENSG00000198886.2</th>\n",
       "      <th>ENSG00000275896.5</th>\n",
       "      <th>ENSG00000163220.10</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000148053.15</th>\n",
       "      <th>ENSG00000134291.11</th>\n",
       "      <th>ENSG00000183578.6</th>\n",
       "      <th>ENSG00000164091.11</th>\n",
       "      <th>ENSG00000173418.11</th>\n",
       "      <th>Name</th>\n",
       "      <th>tissue</th>\n",
       "      <th>SUBJID</th>\n",
       "      <th>SEX</th>\n",
       "      <th>AGE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-0.189282</td>\n",
       "      <td>-0.638404</td>\n",
       "      <td>-0.912095</td>\n",
       "      <td>-1.145540</td>\n",
       "      <td>-0.795922</td>\n",
       "      <td>-0.192404</td>\n",
       "      <td>-1.089657</td>\n",
       "      <td>-1.153692</td>\n",
       "      <td>-0.132456</td>\n",
       "      <td>-0.248378</td>\n",
       "      <td>...</td>\n",
       "      <td>0.062134</td>\n",
       "      <td>-0.841136</td>\n",
       "      <td>0.994222</td>\n",
       "      <td>0.596135</td>\n",
       "      <td>0.676412</td>\n",
       "      <td>GTEX-1117F-0226-SM-5GZZ7</td>\n",
       "      <td>Adipose Tissue</td>\n",
       "      <td>GTEX-1117F</td>\n",
       "      <td>2</td>\n",
       "      <td>60-69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-0.192726</td>\n",
       "      <td>-0.486328</td>\n",
       "      <td>0.055921</td>\n",
       "      <td>0.109423</td>\n",
       "      <td>1.286259</td>\n",
       "      <td>-0.194876</td>\n",
       "      <td>0.890270</td>\n",
       "      <td>-0.001923</td>\n",
       "      <td>-0.129289</td>\n",
       "      <td>-0.248983</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.617946</td>\n",
       "      <td>-1.181526</td>\n",
       "      <td>-0.709254</td>\n",
       "      <td>-0.593598</td>\n",
       "      <td>1.388128</td>\n",
       "      <td>GTEX-1117F-0426-SM-5EGHI</td>\n",
       "      <td>Muscle</td>\n",
       "      <td>GTEX-1117F</td>\n",
       "      <td>2</td>\n",
       "      <td>60-69</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 5005 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   ENSG00000244734.3  ENSG00000210082.2  ENSG00000198804.2  ENSG00000198712.1  \\\n",
       "0          -0.189282          -0.638404          -0.912095          -1.145540   \n",
       "1          -0.192726          -0.486328           0.055921           0.109423   \n",
       "\n",
       "   ENSG00000198938.2  ENSG00000188536.12  ENSG00000198899.2  \\\n",
       "0          -0.795922           -0.192404          -1.089657   \n",
       "1           1.286259           -0.194876           0.890270   \n",
       "\n",
       "   ENSG00000198886.2  ENSG00000275896.5  ENSG00000163220.10  ...  \\\n",
       "0          -1.153692          -0.132456           -0.248378  ...   \n",
       "1          -0.001923          -0.129289           -0.248983  ...   \n",
       "\n",
       "   ENSG00000148053.15  ENSG00000134291.11  ENSG00000183578.6  \\\n",
       "0            0.062134           -0.841136           0.994222   \n",
       "1           -0.617946           -1.181526          -0.709254   \n",
       "\n",
       "   ENSG00000164091.11  ENSG00000173418.11                      Name  \\\n",
       "0            0.596135            0.676412  GTEX-1117F-0226-SM-5GZZ7   \n",
       "1           -0.593598            1.388128  GTEX-1117F-0426-SM-5EGHI   \n",
       "\n",
       "           tissue      SUBJID  SEX    AGE  \n",
       "0  Adipose Tissue  GTEX-1117F    2  60-69  \n",
       "1          Muscle  GTEX-1117F    2  60-69  \n",
       "\n",
       "[2 rows x 5005 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GeneSubj = pd.merge(norm_genes,Subjects[[\"SUBJID\",\"SEX\",\"AGE\"]],on=\"SUBJID\")\n",
    "GeneSubj.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4750abb8",
   "metadata": {},
   "source": [
    "### Subset to Blood and Brain and get gene_cols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8765f177",
   "metadata": {},
   "outputs": [],
   "source": [
    "Gene_BB = GeneSubj[GeneSubj[\"tissue\"].isin([\"Blood\",\"Brain\"])].copy()\n",
    "\n",
    "#Make tissue, and and age categorical variables for the model (could also just do C(tissue) in the model, but this will slightly speed things up across 5k genes)\n",
    "Gene_BB[\"tissue\"] = Gene_BB[\"tissue\"].astype(\"category\")\n",
    "Gene_BB[\"SEX\"] = Gene_BB[\"SEX\"].astype(\"category\")\n",
    "Gene_BB[\"AGE\"] = Gene_BB[\"AGE\"].astype(\"category\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b62a942a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSG00000244734.3</th>\n",
       "      <th>ENSG00000210082.2</th>\n",
       "      <th>ENSG00000198804.2</th>\n",
       "      <th>ENSG00000198712.1</th>\n",
       "      <th>ENSG00000198938.2</th>\n",
       "      <th>ENSG00000188536.12</th>\n",
       "      <th>ENSG00000198899.2</th>\n",
       "      <th>ENSG00000198886.2</th>\n",
       "      <th>ENSG00000275896.5</th>\n",
       "      <th>ENSG00000163220.10</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000148053.15</th>\n",
       "      <th>ENSG00000134291.11</th>\n",
       "      <th>ENSG00000183578.6</th>\n",
       "      <th>ENSG00000164091.11</th>\n",
       "      <th>ENSG00000173418.11</th>\n",
       "      <th>Name</th>\n",
       "      <th>tissue</th>\n",
       "      <th>SUBJID</th>\n",
       "      <th>SEX</th>\n",
       "      <th>AGE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>-0.191377</td>\n",
       "      <td>2.170896</td>\n",
       "      <td>1.860875</td>\n",
       "      <td>1.122935</td>\n",
       "      <td>1.40874</td>\n",
       "      <td>-0.191746</td>\n",
       "      <td>1.151573</td>\n",
       "      <td>1.217473</td>\n",
       "      <td>-0.132327</td>\n",
       "      <td>-0.256958</td>\n",
       "      <td>...</td>\n",
       "      <td>1.150284</td>\n",
       "      <td>-0.142</td>\n",
       "      <td>-0.414157</td>\n",
       "      <td>-1.288017</td>\n",
       "      <td>-0.963595</td>\n",
       "      <td>GTEX-1117F-3226-SM-5N9CT</td>\n",
       "      <td>Brain</td>\n",
       "      <td>GTEX-1117F</td>\n",
       "      <td>2</td>\n",
       "      <td>60-69</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 5005 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    ENSG00000244734.3  ENSG00000210082.2  ENSG00000198804.2  \\\n",
       "11          -0.191377           2.170896           1.860875   \n",
       "\n",
       "    ENSG00000198712.1  ENSG00000198938.2  ENSG00000188536.12  \\\n",
       "11           1.122935            1.40874           -0.191746   \n",
       "\n",
       "    ENSG00000198899.2  ENSG00000198886.2  ENSG00000275896.5  \\\n",
       "11           1.151573           1.217473          -0.132327   \n",
       "\n",
       "    ENSG00000163220.10  ...  ENSG00000148053.15  ENSG00000134291.11  \\\n",
       "11           -0.256958  ...            1.150284              -0.142   \n",
       "\n",
       "    ENSG00000183578.6  ENSG00000164091.11  ENSG00000173418.11  \\\n",
       "11          -0.414157           -1.288017           -0.963595   \n",
       "\n",
       "                        Name  tissue      SUBJID  SEX    AGE  \n",
       "11  GTEX-1117F-3226-SM-5N9CT   Brain  GTEX-1117F    2  60-69  \n",
       "\n",
       "[1 rows x 5005 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Gene_BB.head(1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3b93758",
   "metadata": {},
   "source": [
    "## Linear Modeling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "79eaf67d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 5000/5000 [00:49<00:00, 100.44it/s]\n"
     ]
    }
   ],
   "source": [
    "info=[]\n",
    "for gene in tqdm(gene_cols):\n",
    "    #make linear model and correct for sex and age need to use Q around gene because the names are weird\n",
    "    model = ols(f\"Q('{gene}') ~ tissue + SEX + AGE\", data=Gene_BB).fit()\n",
    "    info.append({\"Gene\": gene,\"coef\": model.params[\"tissue[T.Brain]\"],\"pval\": model.pvalues[\"tissue[T.Brain]\"]})\n",
    "tissue_eval = pd.DataFrame(info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ca67f750",
   "metadata": {},
   "outputs": [],
   "source": [
    "#read in gene descriptions\n",
    "get_description = pd.read_csv(\"Gene_Descriptions.csv\")\n",
    "get_description = dict(zip(get_description[\"Name\"],get_description[\"Desc\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "58f65698",
   "metadata": {},
   "outputs": [],
   "source": [
    "tissue_eval[\"Desc\"] = tissue_eval[\"Gene\"].map(get_description)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "252fc705",
   "metadata": {},
   "source": [
    "#### Top 5 most signficant pvalues"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "87fbe1f9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>coef</th>\n",
       "      <th>pval</th>\n",
       "      <th>Desc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1535</th>\n",
       "      <td>ENSG00000064601.16</td>\n",
       "      <td>-1.183147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>CTSA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3162</th>\n",
       "      <td>ENSG00000105357.15</td>\n",
       "      <td>0.430296</td>\n",
       "      <td>0.0</td>\n",
       "      <td>MYH14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3155</th>\n",
       "      <td>ENSG00000002834.17</td>\n",
       "      <td>-1.666254</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LASP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3154</th>\n",
       "      <td>ENSG00000069275.12</td>\n",
       "      <td>1.681651</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NUCKS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3146</th>\n",
       "      <td>ENSG00000153113.23</td>\n",
       "      <td>-0.587200</td>\n",
       "      <td>0.0</td>\n",
       "      <td>CAST</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Gene      coef  pval    Desc\n",
       "1535  ENSG00000064601.16 -1.183147   0.0    CTSA\n",
       "3162  ENSG00000105357.15  0.430296   0.0   MYH14\n",
       "3155  ENSG00000002834.17 -1.666254   0.0   LASP1\n",
       "3154  ENSG00000069275.12  1.681651   0.0  NUCKS1\n",
       "3146  ENSG00000153113.23 -0.587200   0.0    CAST"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tissue_eval.sort_values(\"pval\").head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "06294694",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of underflow pvalues: 967\n",
      "Number of not signficant genes: 294\n"
     ]
    }
   ],
   "source": [
    "print(\"Number of underflow pvalues:\",tissue_eval[tissue_eval[\"pval\"]==0][\"Gene\"].count())\n",
    "print(\"Number of not signficant genes:\",tissue_eval[tissue_eval[\"pval\"]>.05][\"Gene\"].count())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb445998",
   "metadata": {},
   "source": [
    "#### Top 5 most highest effect genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "0e68432e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>coef</th>\n",
       "      <th>pval</th>\n",
       "      <th>Desc</th>\n",
       "      <th>abs_coef</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>638</th>\n",
       "      <td>ENSG00000128340.14</td>\n",
       "      <td>-3.965246</td>\n",
       "      <td>0.0</td>\n",
       "      <td>RAC2</td>\n",
       "      <td>3.965246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>493</th>\n",
       "      <td>ENSG00000136167.13</td>\n",
       "      <td>-3.914198</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LCP1</td>\n",
       "      <td>3.914198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1255</th>\n",
       "      <td>ENSG00000147168.12</td>\n",
       "      <td>-3.862512</td>\n",
       "      <td>0.0</td>\n",
       "      <td>IL2RG</td>\n",
       "      <td>3.862512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1777</th>\n",
       "      <td>ENSG00000130592.15</td>\n",
       "      <td>-3.860039</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LSP1</td>\n",
       "      <td>3.860039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>449</th>\n",
       "      <td>ENSG00000162511.7</td>\n",
       "      <td>-3.834085</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LAPTM5</td>\n",
       "      <td>3.834085</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Gene      coef  pval    Desc  abs_coef\n",
       "638   ENSG00000128340.14 -3.965246   0.0    RAC2  3.965246\n",
       "493   ENSG00000136167.13 -3.914198   0.0    LCP1  3.914198\n",
       "1255  ENSG00000147168.12 -3.862512   0.0   IL2RG  3.862512\n",
       "1777  ENSG00000130592.15 -3.860039   0.0    LSP1  3.860039\n",
       "449    ENSG00000162511.7 -3.834085   0.0  LAPTM5  3.834085"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tissue_eval[\"abs_coef\"]=tissue_eval[\"coef\"].abs()\n",
    "tissue_eval.sort_values(\"abs_coef\",ascending=False).head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "742fe766",
   "metadata": {},
   "source": [
    "Interpretation:\n",
    "\n",
    "There is not overlap between the the most significant genes and the highest effect (coef). Two reasons:\n",
    "\n",
    "1) The pvalues are extremely tiny and for both sets and underflow to 0 (would be reported as <0.0001), so sorting a bunch of 0s is useless. There are much greater than only 5 values that underflow to 0 (967 genes) making top 5 pvalues meaningless. \n",
    "\n",
    "2) Pvalue and abs coef are different. Pvalue controls for both noise and sample size to see the strength of dif between brain and blood for that gene. The coef simply how big the gene difference is between brain and blood ignoring noise and sample size.\n",
    "\n",
    "Since Brain and Blood are extremely different tissues it is not surprising to see many very sig dif genes. In fact, only 295 genes of 5000 are NOT significantly different. Thus, it is unsurprising that there is not overlap in just taking the top 5 smallest pvalues because it is essentially a random sample of 5 of the 967 most significant genes (where the pval underflows to 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82836a7a",
   "metadata": {},
   "source": [
    "## Logistic Modeling\n",
    "\n",
    "Predict if gene expression can predict sex in blood only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "736284d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "Gene_Blood = Gene_BB[Gene_BB[\"tissue\"]==\"Blood\"].copy()\n",
    "#make sex a binary numerical variable 0,1\n",
    "Gene_Blood[\"SEX\"] = (Gene_Blood[\"SEX\"] == 2).astype(\"int\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a89ddb97",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 5000/5000 [00:24<00:00, 205.76it/s]\n"
     ]
    }
   ],
   "source": [
    "warnings.filterwarnings(\"ignore\", category=ConvergenceWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=RuntimeWarning)\n",
    "\n",
    "info_log=[]\n",
    "for gene in tqdm(gene_cols):\n",
    "    log_model = logit(f\"SEX ~ Q('{gene}')\", data=Gene_Blood).fit(disp=0)\n",
    "    info_log.append({\"Gene\": gene,\"coef\": log_model.params.iloc[1],\"pval\": log_model.pvalues.iloc[1]})\n",
    "    \n",
    "log_SEX_eval = pd.DataFrame(info_log)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "1ebca4e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "log_SEX_eval[\"Desc\"] = log_SEX_eval[\"Gene\"].map(get_description)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02b5360c",
   "metadata": {},
   "source": [
    "#### Top 5 most signficant pvalues logistic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "d5990488",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>coef</th>\n",
       "      <th>pval</th>\n",
       "      <th>Desc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3386</th>\n",
       "      <td>ENSG00000229807.10</td>\n",
       "      <td>482.625332</td>\n",
       "      <td>4.069779e-20</td>\n",
       "      <td>XIST</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349</th>\n",
       "      <td>ENSG00000198034.10</td>\n",
       "      <td>0.425305</td>\n",
       "      <td>6.965764e-12</td>\n",
       "      <td>RPS4X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2638</th>\n",
       "      <td>ENSG00000215301.9</td>\n",
       "      <td>0.491742</td>\n",
       "      <td>4.678563e-11</td>\n",
       "      <td>DDX3X</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4199</th>\n",
       "      <td>ENSG00000067048.16</td>\n",
       "      <td>-187.174051</td>\n",
       "      <td>1.517101e-09</td>\n",
       "      <td>DDX3Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3021</th>\n",
       "      <td>ENSG00000130985.16</td>\n",
       "      <td>0.379586</td>\n",
       "      <td>1.583458e-08</td>\n",
       "      <td>UBA1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Gene        coef          pval   Desc\n",
       "3386  ENSG00000229807.10  482.625332  4.069779e-20   XIST\n",
       "349   ENSG00000198034.10    0.425305  6.965764e-12  RPS4X\n",
       "2638   ENSG00000215301.9    0.491742  4.678563e-11  DDX3X\n",
       "4199  ENSG00000067048.16 -187.174051  1.517101e-09  DDX3Y\n",
       "3021  ENSG00000130985.16    0.379586  1.583458e-08   UBA1"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "log_SEX_eval.sort_values(\"pval\").head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "687bef02",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of underflow pvalues: 0\n",
      "Number of not signficant genes: 541\n",
      "Number of not signficant genes: 4459\n"
     ]
    }
   ],
   "source": [
    "print(\"Number of underflow pvalues:\",log_SEX_eval[log_SEX_eval[\"pval\"]==0][\"Gene\"].count())\n",
    "print(\"Number of not signficant genes:\",log_SEX_eval[log_SEX_eval[\"pval\"]<.05][\"Gene\"].count())\n",
    "print(\"Number of not signficant genes:\",log_SEX_eval[log_SEX_eval[\"pval\"]>.05][\"Gene\"].count())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d26d983b",
   "metadata": {},
   "source": [
    "Notes: Most significant gene is XIST which makes complete sense because this gene silences one X chromosome in females. This gene would not be expressed in males. The next three genes have related expression to the X and Y chromosomes (X or Y after the gene show this and confirmed in gene cards) DDX3Y is strictly Y linked, so this also makes sense."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71cc459c",
   "metadata": {},
   "source": [
    "#### Top 5 most highest effect genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "c1f0100a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>coef</th>\n",
       "      <th>pval</th>\n",
       "      <th>Desc</th>\n",
       "      <th>abs_coef</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4751</th>\n",
       "      <td>ENSG00000012817.15</td>\n",
       "      <td>-3962.612072</td>\n",
       "      <td>0.994726</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>3962.612072</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1159</th>\n",
       "      <td>ENSG00000129824.15</td>\n",
       "      <td>-1422.133879</td>\n",
       "      <td>0.999319</td>\n",
       "      <td>RPS4Y1</td>\n",
       "      <td>1422.133879</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>723</th>\n",
       "      <td>ENSG00000135222.6</td>\n",
       "      <td>1307.963203</td>\n",
       "      <td>0.119342</td>\n",
       "      <td>CSN2</td>\n",
       "      <td>1307.963203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>770</th>\n",
       "      <td>ENSG00000124157.6</td>\n",
       "      <td>-1034.619007</td>\n",
       "      <td>0.120261</td>\n",
       "      <td>SEMG2</td>\n",
       "      <td>1034.619007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>406</th>\n",
       "      <td>ENSG00000171209.3</td>\n",
       "      <td>925.322653</td>\n",
       "      <td>0.092940</td>\n",
       "      <td>CSN3</td>\n",
       "      <td>925.322653</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Gene         coef      pval    Desc     abs_coef\n",
       "4751  ENSG00000012817.15 -3962.612072  0.994726   KDM5D  3962.612072\n",
       "1159  ENSG00000129824.15 -1422.133879  0.999319  RPS4Y1  1422.133879\n",
       "723    ENSG00000135222.6  1307.963203  0.119342    CSN2  1307.963203\n",
       "770    ENSG00000124157.6 -1034.619007  0.120261   SEMG2  1034.619007\n",
       "406    ENSG00000171209.3   925.322653  0.092940    CSN3   925.322653"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "log_SEX_eval[\"abs_coef\"]=log_SEX_eval[\"coef\"].abs()\n",
    "log_SEX_eval.sort_values(\"abs_coef\",ascending=False).head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd8a44c1",
   "metadata": {},
   "source": [
    "Interpretation:\n",
    "\n",
    "To revisit what I stated in the prior section about how pvalue and coef are different: pvalue controls for both noise and sample size to see the strength of dif for the gene. The coef simply how big the gene difference is between male and female ignoring how noisy it may be. This is likely why none of the biggest coefs are significant. "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Scienv_LT1",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
