{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a09d7104",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from sklearn.cluster import KMeans\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.feature_selection import f_classif"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4472fec0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>ENSG00000244734.3</th>\n",
       "      <th>ENSG00000210082.2</th>\n",
       "      <th>ENSG00000198804.2</th>\n",
       "      <th>ENSG00000198712.1</th>\n",
       "      <th>ENSG00000198938.2</th>\n",
       "      <th>ENSG00000188536.12</th>\n",
       "      <th>ENSG00000198899.2</th>\n",
       "      <th>ENSG00000198886.2</th>\n",
       "      <th>ENSG00000275896.5</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000176749.8</th>\n",
       "      <th>ENSG00000241973.10</th>\n",
       "      <th>ENSG00000118689.14</th>\n",
       "      <th>ENSG00000167680.15</th>\n",
       "      <th>ENSG00000148053.15</th>\n",
       "      <th>ENSG00000134291.11</th>\n",
       "      <th>ENSG00000183578.6</th>\n",
       "      <th>ENSG00000164091.11</th>\n",
       "      <th>ENSG00000173418.11</th>\n",
       "      <th>tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GTEX-1117F-0226-SM-5GZZ7</td>\n",
       "      <td>452.7</td>\n",
       "      <td>6310.0</td>\n",
       "      <td>10790.0</td>\n",
       "      <td>11720.0</td>\n",
       "      <td>19890.0</td>\n",
       "      <td>102.40</td>\n",
       "      <td>13880.0</td>\n",
       "      <td>12400.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0290</td>\n",
       "      <td>31.85</td>\n",
       "      <td>62.17</td>\n",
       "      <td>24.530</td>\n",
       "      <td>18.2900</td>\n",
       "      <td>17.720</td>\n",
       "      <td>45.8000</td>\n",
       "      <td>76.38</td>\n",
       "      <td>61.47</td>\n",
       "      <td>Adipose Tissue</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GTEX-1117F-0426-SM-5EGHI</td>\n",
       "      <td>225.7</td>\n",
       "      <td>10200.0</td>\n",
       "      <td>33610.0</td>\n",
       "      <td>37500.0</td>\n",
       "      <td>62560.0</td>\n",
       "      <td>52.01</td>\n",
       "      <td>51690.0</td>\n",
       "      <td>34030.0</td>\n",
       "      <td>46.660</td>\n",
       "      <td>...</td>\n",
       "      <td>1.6640</td>\n",
       "      <td>14.18</td>\n",
       "      <td>110.90</td>\n",
       "      <td>9.736</td>\n",
       "      <td>0.2841</td>\n",
       "      <td>8.708</td>\n",
       "      <td>0.7015</td>\n",
       "      <td>44.89</td>\n",
       "      <td>80.30</td>\n",
       "      <td>Muscle</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GTEX-1117F-0526-SM-5EGHJ</td>\n",
       "      <td>269.6</td>\n",
       "      <td>6111.0</td>\n",
       "      <td>9689.0</td>\n",
       "      <td>12250.0</td>\n",
       "      <td>19450.0</td>\n",
       "      <td>62.95</td>\n",
       "      <td>16270.0</td>\n",
       "      <td>13820.0</td>\n",
       "      <td>159.200</td>\n",
       "      <td>...</td>\n",
       "      <td>1.6410</td>\n",
       "      <td>30.01</td>\n",
       "      <td>73.15</td>\n",
       "      <td>33.440</td>\n",
       "      <td>16.6400</td>\n",
       "      <td>33.610</td>\n",
       "      <td>77.1700</td>\n",
       "      <td>62.28</td>\n",
       "      <td>46.12</td>\n",
       "      <td>Blood Vessel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GTEX-1117F-0626-SM-5N9CS</td>\n",
       "      <td>5272.0</td>\n",
       "      <td>11990.0</td>\n",
       "      <td>4269.0</td>\n",
       "      <td>10630.0</td>\n",
       "      <td>16440.0</td>\n",
       "      <td>1323.00</td>\n",
       "      <td>16350.0</td>\n",
       "      <td>11990.0</td>\n",
       "      <td>3.747</td>\n",
       "      <td>...</td>\n",
       "      <td>1.8950</td>\n",
       "      <td>39.19</td>\n",
       "      <td>60.41</td>\n",
       "      <td>29.950</td>\n",
       "      <td>21.3700</td>\n",
       "      <td>58.980</td>\n",
       "      <td>38.1100</td>\n",
       "      <td>74.41</td>\n",
       "      <td>51.21</td>\n",
       "      <td>Blood Vessel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GTEX-1117F-0726-SM-5GIEN</td>\n",
       "      <td>1617.0</td>\n",
       "      <td>40440.0</td>\n",
       "      <td>56700.0</td>\n",
       "      <td>39220.0</td>\n",
       "      <td>64660.0</td>\n",
       "      <td>403.50</td>\n",
       "      <td>82550.0</td>\n",
       "      <td>69350.0</td>\n",
       "      <td>9.192</td>\n",
       "      <td>...</td>\n",
       "      <td>0.6289</td>\n",
       "      <td>15.62</td>\n",
       "      <td>20.16</td>\n",
       "      <td>14.000</td>\n",
       "      <td>4.5110</td>\n",
       "      <td>23.490</td>\n",
       "      <td>5.6750</td>\n",
       "      <td>31.22</td>\n",
       "      <td>22.02</td>\n",
       "      <td>Heart</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 5002 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Name  ENSG00000244734.3  ENSG00000210082.2  \\\n",
       "0  GTEX-1117F-0226-SM-5GZZ7              452.7             6310.0   \n",
       "1  GTEX-1117F-0426-SM-5EGHI              225.7            10200.0   \n",
       "2  GTEX-1117F-0526-SM-5EGHJ              269.6             6111.0   \n",
       "3  GTEX-1117F-0626-SM-5N9CS             5272.0            11990.0   \n",
       "4  GTEX-1117F-0726-SM-5GIEN             1617.0            40440.0   \n",
       "\n",
       "   ENSG00000198804.2  ENSG00000198712.1  ENSG00000198938.2  \\\n",
       "0            10790.0            11720.0            19890.0   \n",
       "1            33610.0            37500.0            62560.0   \n",
       "2             9689.0            12250.0            19450.0   \n",
       "3             4269.0            10630.0            16440.0   \n",
       "4            56700.0            39220.0            64660.0   \n",
       "\n",
       "   ENSG00000188536.12  ENSG00000198899.2  ENSG00000198886.2  \\\n",
       "0              102.40            13880.0            12400.0   \n",
       "1               52.01            51690.0            34030.0   \n",
       "2               62.95            16270.0            13820.0   \n",
       "3             1323.00            16350.0            11990.0   \n",
       "4              403.50            82550.0            69350.0   \n",
       "\n",
       "   ENSG00000275896.5  ...  ENSG00000176749.8  ENSG00000241973.10  \\\n",
       "0              0.000  ...             2.0290               31.85   \n",
       "1             46.660  ...             1.6640               14.18   \n",
       "2            159.200  ...             1.6410               30.01   \n",
       "3              3.747  ...             1.8950               39.19   \n",
       "4              9.192  ...             0.6289               15.62   \n",
       "\n",
       "   ENSG00000118689.14  ENSG00000167680.15  ENSG00000148053.15  \\\n",
       "0               62.17              24.530             18.2900   \n",
       "1              110.90               9.736              0.2841   \n",
       "2               73.15              33.440             16.6400   \n",
       "3               60.41              29.950             21.3700   \n",
       "4               20.16              14.000              4.5110   \n",
       "\n",
       "   ENSG00000134291.11  ENSG00000183578.6  ENSG00000164091.11  \\\n",
       "0              17.720            45.8000               76.38   \n",
       "1               8.708             0.7015               44.89   \n",
       "2              33.610            77.1700               62.28   \n",
       "3              58.980            38.1100               74.41   \n",
       "4              23.490             5.6750               31.22   \n",
       "\n",
       "   ENSG00000173418.11          tissue  \n",
       "0               61.47  Adipose Tissue  \n",
       "1               80.30          Muscle  \n",
       "2               46.12    Blood Vessel  \n",
       "3               51.21    Blood Vessel  \n",
       "4               22.02           Heart  \n",
       "\n",
       "[5 rows x 5002 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_feats = pd.read_csv(\"ClusterDataset.csv\")\n",
    "gene_feats.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90b77de2",
   "metadata": {},
   "source": [
    "### Remove gene a label columns. Normalize Data useing standard scaler"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "c28a2049",
   "metadata": {},
   "outputs": [],
   "source": [
    "labels = gene_feats[\"tissue\"].to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3b9cb36",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Get labels\n",
    "labels = gene_feats[\"tissue\"].to_list()\n",
    "#Remove gene column and features column convert to numpy array\n",
    "gene_array = gene_feats.iloc[:,1:-1].to_numpy()\n",
    "norm_genes = StandardScaler().fit_transform(gene_array)\n",
    "gene_names = gene_feats.columns[1:-1] #get only gene names remove Name and labels columns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f646ed58",
   "metadata": {},
   "source": [
    "### Based on ClusterAssignment \n",
    "\n",
    "There are no particularly strong clusters from internal metrics\n",
    "\n",
    "Use cluster of 10 since we know there are 10 tissues and evaluate if there are any genes driving the clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc27da9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Obtain clusters\n",
    "km = KMeans(n_clusters=10, n_init=\"auto\", random_state=0,algorithm='elkan')\n",
    "labels_km = km.fit_predict(norm_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6900125f",
   "metadata": {},
   "source": [
    "### Use one way ANOVA for each gene to 10 Kmeans Clusters\n",
    "\n",
    "One way ANOVA was chosen because it separates between group vs. within group variation. I have many groups (either the clustering or the tissue category) and genes are continuouis variables.\n",
    "\n",
    "\n",
    "\n",
    "If a gene has a large F value, then it varies a lot across clusters. \n",
    "\n",
    "Sorting by F value is the same as sorting by pvalue because large F = smaller pvalue.\n",
    "\n",
    "This would suggest it is a driver of cluster separation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "1b390488",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ENSG00000244734.3     5.778444e+08\n",
       "ENSG00000210082.2     1.473324e+08\n",
       "ENSG00000198804.2     8.912685e+07\n",
       "ENSG00000198712.1     7.237725e+07\n",
       "ENSG00000198938.2     6.118706e+07\n",
       "ENSG00000198899.2     5.953502e+07\n",
       "ENSG00000188536.12    5.532686e+07\n",
       "ENSG00000198886.2     5.136573e+07\n",
       "ENSG00000171401.14    3.591510e+07\n",
       "ENSG00000186395.7     2.897677e+07\n",
       "ENSG00000198888.2     2.856604e+07\n",
       "ENSG00000163220.10    2.704641e+07\n",
       "ENSG00000275896.5     2.626163e+07\n",
       "ENSG00000198763.3     2.420453e+07\n",
       "ENSG00000198727.2     2.117964e+07\n",
       "ENSG00000259384.6     1.921000e+07\n",
       "ENSG00000204983.13    1.852833e+07\n",
       "ENSG00000211459.2     1.770776e+07\n",
       "ENSG00000143632.14    1.703461e+07\n",
       "ENSG00000228253.1     1.607809e+07\n",
       "dtype: float64"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f_vals, p_vals = f_classif(norm_genes,labels_km)\n",
    "f_series = pd.Series(f_vals, index=gene_names)\n",
    "top_n = 20\n",
    "top_genes_clusters = f_series.sort_values(ascending=False).head(top_n)\n",
    "top_genes_clusters"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3eac067",
   "metadata": {},
   "source": [
    "### Get Gene descriptions for interpretation of findings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6315eab",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Download Data here: https://gtexportal.org/home/downloads/adult-gtex/bulk_tissue_expression\n",
    "Gene = pd.read_csv(\"GTEx_Analysis_20170605_v8_RNASeQCv1.1.9_gene_tpm.gct\",sep =\"\\t\",skiprows=2,engine=\"c\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "b7ff1c65",
   "metadata": {},
   "outputs": [],
   "source": [
    "get_description = dict(zip(Gene[\"Name\"],Gene[\"Description\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "7af5e144",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ENSG00000244734.3': 'HBB',\n",
       " 'ENSG00000210082.2': 'MT-RNR2',\n",
       " 'ENSG00000198804.2': 'MT-CO1',\n",
       " 'ENSG00000198712.1': 'MT-CO2',\n",
       " 'ENSG00000198938.2': 'MT-CO3',\n",
       " 'ENSG00000198899.2': 'MT-ATP6',\n",
       " 'ENSG00000188536.12': 'HBA2',\n",
       " 'ENSG00000198886.2': 'MT-ND4',\n",
       " 'ENSG00000171401.14': 'KRT13',\n",
       " 'ENSG00000186395.7': 'KRT10',\n",
       " 'ENSG00000198888.2': 'MT-ND1',\n",
       " 'ENSG00000163220.10': 'S100A9',\n",
       " 'ENSG00000275896.5': 'PRSS2',\n",
       " 'ENSG00000198763.3': 'MT-ND2',\n",
       " 'ENSG00000198727.2': 'MT-CYB',\n",
       " 'ENSG00000259384.6': 'GH1',\n",
       " 'ENSG00000204983.13': 'PRSS1',\n",
       " 'ENSG00000211459.2': 'MT-RNR1',\n",
       " 'ENSG00000143632.14': 'ACTA1',\n",
       " 'ENSG00000228253.1': 'MT-ATP8'}"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t20genes={}\n",
    "for g in top_genes_clusters.index:\n",
    "    t20genes[g] = get_description[g]\n",
    "t20genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49a6c8fb",
   "metadata": {},
   "source": [
    "### Interpretation\n",
    "\n",
    "Top gene is hemoglobin beta which makes sense especially since this likely clusters blood tissue.\n",
    "\n",
    "Hemoglobin alpha 2 is also in the top genes, likely for a similar reason.\n",
    "\n",
    "The majority of top genes are mitochondrial related which may be related to differences in metabolic demand between tissues. For example, red blood cells do not have mitochondria so lacking this expression could drive that group. \n",
    "\n",
    "Another reason is this may just reflect differences in the stress each sample was in when it was being processed. I would expect more degraded samples to have higher mitochondrial reads. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e90045b",
   "metadata": {},
   "source": [
    "### Use one way ANOVA for each Tissue label "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "06951441",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ENSG00000167476.10    8154.952866\n",
       "ENSG00000138347.15    6755.844628\n",
       "ENSG00000169862.18    6404.067451\n",
       "ENSG00000079393.20    5758.511665\n",
       "ENSG00000058056.8     5641.304562\n",
       "ENSG00000178104.19    5529.093124\n",
       "ENSG00000104369.4     5303.005845\n",
       "ENSG00000154358.20    5281.136270\n",
       "ENSG00000143028.8     5276.027456\n",
       "ENSG00000133454.15    5081.811567\n",
       "ENSG00000133315.10    4977.701651\n",
       "ENSG00000164776.9     4925.772189\n",
       "ENSG00000163126.14    4786.690601\n",
       "ENSG00000229444.1     4774.072717\n",
       "ENSG00000132692.18    4711.212088\n",
       "ENSG00000174939.10    4438.099903\n",
       "ENSG00000143164.15    4391.094053\n",
       "ENSG00000087258.14    4290.132432\n",
       "ENSG00000163380.15    4090.750569\n",
       "ENSG00000081248.10    4075.215437\n",
       "dtype: float64"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f_vals, p_vals = f_classif(norm_genes,labels)\n",
    "f_series = pd.Series(f_vals, index=gene_names)\n",
    "top_n = 20\n",
    "top_genes_clusters = f_series.sort_values(ascending=False).head(top_n)\n",
    "top_genes_clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "749ac6f0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ENSG00000167476.10': 'JSRP1',\n",
       " 'ENSG00000138347.15': 'MYPN',\n",
       " 'ENSG00000169862.18': 'CTNND2',\n",
       " 'ENSG00000079393.20': 'DUSP13',\n",
       " 'ENSG00000058056.8': 'USP13',\n",
       " 'ENSG00000178104.19': 'PDE4DIP',\n",
       " 'ENSG00000104369.4': 'JPH1',\n",
       " 'ENSG00000154358.20': 'OBSCN',\n",
       " 'ENSG00000143028.8': 'SYPL2',\n",
       " 'ENSG00000133454.15': 'MYO18B',\n",
       " 'ENSG00000133315.10': 'MACROD1',\n",
       " 'ENSG00000164776.9': 'PHKG1',\n",
       " 'ENSG00000163126.14': 'ANKRD23',\n",
       " 'ENSG00000229444.1': 'RP11-184I16.4',\n",
       " 'ENSG00000132692.18': 'BCAN',\n",
       " 'ENSG00000174939.10': 'ASPHD1',\n",
       " 'ENSG00000143164.15': 'DCAF6',\n",
       " 'ENSG00000087258.14': 'GNAO1',\n",
       " 'ENSG00000163380.15': 'LMOD3',\n",
       " 'ENSG00000081248.10': 'CACNA1S'}"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t20genes_tissue={}\n",
    "for g in top_genes_clusters.index:\n",
    "    t20genes_tissue[g] = get_description[g]\n",
    "t20genes_tissue"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34a6b60c",
   "metadata": {},
   "source": [
    "### Interpretation\n",
    "\n",
    "Spot checking on gene cards these are generally genes associated either with skeletal muscle such as JSRP1 (sarcoplasmic reticulum in skeletal muscle), MYPN (Z-disc/I-band sarcomeric protein) or neurons such as CTNND2 (synaptic protein in neurons), BCAN (component of CNS extracellular matrix)."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Scienv_LT1",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
